src/freenet/support/Waiter.java

```diff
@@ -0,0 +1,23 @@
+package freenet.support;
+
+/** Waits for some other thread to signal to proceed. */
+public class Waiter {
```

Comment:
Semaphore is over-complex, to the point of creating new bugs. CountDownLatch looks more promising. Will fix.